### PR TITLE
perf(nuxt, kit, vite): Enhance Performance with Concurrent Execution of Async Functions 🏎

### PR DESCRIPTION
**Pull request by sanjaiyan-dev**
### 📚 Description

Hello, this pull request aims to provide a minor performance boost by concurrently executing asynchronous functions using `Promise.all`. This approach enhances the utilization of hardware resources, leading to improved efficiency and effectiveness. Thank you for your consideration.

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/kit/src/module/define.ts [hunk 1]
```diff
@@ -143,10 +143,12 @@ function nuxt2Shims (nuxt: Nuxt) {
         templateVars: templates.templateVars
       }
     }
+    const promises: Promise<void>[] = [];
     for await (const template of virtualTemplates) {
       const contents = await compileTemplate({ ...template, src: '' }, context)
       await fsp.mkdir(dirname(template.dst), { recursive: true })
```
Comment (GalacticHypernova) [inline packages/kit/src/module/define.ts:149]: We can probably place this in another Promise.all that would run before the Promise.all(promises) if we want to really speed it up.

**Comment by GalacticHypernova** [comment]
image

Appears to be a bit mixed, but I do agree that for the initial installation it would make a difference

**Comment by sanjaiyan-dev** [comment]
> image
> 
> Appears to be a bit mixed with the current method being sometimes faster, but I do agree that for the initial installation it would make a difference Also, due to the similarity this has in nature to #24771 , is it possible to perhaps modify that changelog to include this? I don't mind if all credit would go to sanjaiyan and my commit would get overwritten, or if there's a way to combine them (commit1 + commit2) in the changelog, since the schema update also did this, so it would only make sense to group it in some way.

Hi @GalacticHypernova ,

Thank you for acknowledging the credit, but your contribution is more significant.
I still wonder why concurrent processing experienced a small performance drop ‽

**Comment by GalacticHypernova** [comment]
> I still wonder why concurrent processing experienced a small performance drop ‽

I wouldn't count on it, it seems to give out very inaccurate reports.

**Comment by sanjaiyan-dev** [comment]
> I wouldn't count on it, it seems to give out very inaccurate reports.

Understood, thank you.

**Review by danielroe — CHANGES_REQUESTED** [review]
we did a small benchmark on this but couldn't find that making FS access parallel was beneficial as writing to the file system generally has to be done sequentially anyway. (Though compiling the templates in parallel seems useful.)

The bit you've changed here is for Nuxt 2. The way we improved this for Nuxt 3 here:

https://github.com/nuxt/nuxt/blob/5877e11c89fa3f6cf5699ca1a96d1f8b67605f8f/packages/nuxt/src/core/app.ts#L43-L75.

I'd welcome a PR that ports that improvement to Nuxt 2 within `packages/kit/src/module/define.ts`.

(We can skip the change in `packages/nuxt/src/core/nitro.ts`.)

Marking as draft just for my own triage - feel free to unmark as draft whenever it's ready to review 🙏

**Pull request closed without merging**